include/boost/compute/random/threefry_engine.hpp

```diff
@@ -68,14 +77,81 @@ class threefry_engine
     {
     }
 
-private:
+    /// Seeds the random number generator with \p value.
+    ///
+    /// \param value seed value for the random-number generator
+    /// \param queue command queue to perform the operation
+    ///
+    /// If no seed value is provided, \c default_seed is used.
+    void seed(ulong_ value, command_queue &queue)
+    {
+        (void) queue;
+        m_key[0] = static_cast<uint_>(value);
+        m_key[1] = static_cast<uint_>(value >> 32);
+        // Reset counter
+        m_counter[0] = 0;
+        m_counter[1] = 0;
+    }
+
+    /// \overload
+    void seed(command_queue &queue)
+    {
+        seed(default_seed, queue);
+    }
+
+    /// Generates random numbers and stores them to the range [\p first, \p last).
+    template<class OutputIterator>
+    void generate(OutputIterator first, OutputIterator last, command_queue &queue)
+    {
+        const size_t size = detail::iterator_range_size(first, last);
+
+        kernel fill_kernel(m_program, "fill");
+        fill_kernel.set_arg(0, first.get_buffer());
+        fill_kernel.set_arg(1, static_cast<const uint_>(size));
+        fill_kernel.set_arg(2, m_key);
+        fill_kernel.set_arg(3, m_counter);
+
+        queue.enqueue_1d_range_kernel(fill_kernel, 0, (size + 1)/2, 0);
+
+        discard(size, queue);
+    }
+
     /// \internal_
+    void generate(discard_iterator first, discard_iterator last, command_queue &queue)
+    {
+        (void) queue;
+        ulong_ offset = std::distance(first, last);
+
+        uint_ lo = static_cast<uint_>(offset);
+        uint_ hi = static_cast<uint_>(offset >> 32);
+
+        m_counter[0] += lo;
+        m_counter[1] += hi + (m_counter[0] < lo ? 1 : 0);
```

Comment:
consider making counter a 64bit integer just for making this logic simpler. This also allows to make counter an atomic for thread safety.

Reply:
I can make it `ulong_` (and the `m_key` too), but I'm not sure about making the counter an atomic value. Other Boost.Compute RNGs do not promise thread safety (just like RNGs in C++ Standard Library).